File: runtime/src/main/java/com/epam/drill/auto/test/agent/penetration/Strategy.java
package com.epam.drill.auto.test.agent.penetration;

import javassist.CannotCompileException;
import javassist.CtClass;
import javassist.NotFoundException;

import java.io.IOException;

public abstract class Strategy {
    public abstract String id();

    public abstract boolean permit(CtClass ctClass);

    public abstract byte[] instrument(CtClass ctClass) throws CannotCompileException, IOException, NotFoundException;

}

File: src/runtimeMain/kotlin/com/epam/drill/test/agent/instrumentation/Strategy.kt
/**
 * Copyright 2020 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.epam.drill.test.agent.instrumentation

import com.epam.drill.logger.*
import javassist.ClassPool
import javassist.CtClass
import java.security.ProtectionDomain

abstract class Strategy {
    val logger = Logging.logger { this::class.java.name }

    abstract fun permit(ctClass: CtClass): Boolean

    abstract fun instrument(
        ctClass: CtClass,
        pool: ClassPool,
        classLoader: ClassLoader?,
        protectionDomain: ProtectionDomain?
    ): ByteArray?
}
